lib/src/rules/function_expression_with_incorrect_types.dart

```diff
@@ -0,0 +1,110 @@
+
+import 'package:analyzer/dart/ast/ast.dart';
+import 'package:analyzer/dart/ast/visitor.dart';
+import 'package:analyzer/dart/element/nullability_suffix.dart';
+import 'package:analyzer/dart/element/type.dart';
+import 'package:collection/collection.dart';
+
+import '../analyzer.dart';
+
+const _desc =
+    r'Use the parameter types expected by the target in function expression.';
+
+const _details = r'''
+
+Use the parameter types expected by the target in function expression.
+
+**BAD:**
+```
+f(void Function(int a) p) {}
+
+f((int? a) {});
+f((num a) {});
+```
+
+**GOOD:**
+```
+f(void Function(int a) p) {}
+
+f((a) {});
+f((int a) {});
+```
+
+''';
+
+class FunctionExpressionWithIncorrectTypes extends LintRule
+    implements NodeLintRule {
+  FunctionExpressionWithIncorrectTypes()
+      : super(
+            name: 'function_expression_with_incorrect_types',
+            description: _desc,
+            details: _details,
+            group: Group.style);
+
+  @override
+  void registerNodeProcessors(
+      NodeLintRegistry registry, LinterContext context) {
+    final visitor = _Visitor(this);
+    registry.addFunctionExpression(this, visitor);
+  }
+}
+
+class _Visitor extends SimpleAstVisitor<void> {
+  final LintRule rule;
+
+  _Visitor(this.rule);
+
+  @override
+  void visitFunctionExpression(FunctionExpression node) {
+    DartType? expectedType;
+    var parent = node.unParenthesized.parent;
+    if (parent is AssignmentExpression) {
+      expectedType = parent.writeType;
+    } else {
+      expectedType = node.unParenthesized.staticParameterElement?.type;
+    }
+
+    // exit if the case is not covered
+    if (expectedType == null || expectedType is! FunctionType) return;
+
+    // check type of each parameters
+    var currentType = node.staticType! as FunctionType;
+    var currentParams = currentType.parameters;
+    var expectedParams = expectedType.parameters;
+    for (var currentParam in currentParams) {
+      // if param is mutated, no lint
+      if (node.body.isPotentiallyMutatedInScope(currentParam) ||
+          node.body.isPotentiallyMutatedInClosure(currentParam)) {
+        continue;
+      }
+
+      // if param is not found on target, no lint
+      var expectedParam =
+          expectedParams.where((e) => e.name == currentParam.name).firstOrNull;
+      if (expectedParam == null) {
+        continue;
+      }
+
+      var currentType = currentParam.type;
+      var expectedType = expectedParam.type;
+
+      // if Object? is used instead of dynamic, no lint
+      if (expectedType.isDynamic &&
+          currentType.isDartCoreObject &&
+          currentType.nullabilitySuffix != NullabilitySuffix.none) {
+        continue;
+      }
+
+      // if type are not consistent, lint!
+      if (currentParam.isRequiredPositional && currentType != expectedType ||
+          currentParam.isOptional &&
+              currentType.element != expectedType.element) {
```

Comment:
This condition doesn't cover the case of a required named parameter. It seems like it ought to.